Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2023 Adrian Wüthrich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adrian Wüthrich
+-/
+import Mathlib.Combinatorics.SimpleGraph.AdjMatrix
+import Mathlib.Combinatorics.SimpleGraph.IncMatrix
+import Mathlib.LinearAlgebra.Matrix.PosDef
+import Mathlib.LinearAlgebra.Matrix.IsDiag
+
+/-!
+# Laplacian Matrix
+
+This module defines the Laplacian matrix of a graph, and proves some of its elementary properties.
+
+## Main definitions & Results
+
+* `SimpleGraph.degMatrix`: The degree matrix of a simple graph
+* `SimpleGraph.lapMatrix`: The Laplacian matrix of a simple graph, defined as the difference
+  between the degree matrix and the adjacency matrix.
+* `isPosSemidef_lapMatrix`: The Laplacian matrix is positive semidefinite.
+* `rank_ker_lapMatrix_eq_card_ConnectedComponent`: The number of connected components in `G` is
+  the dimension of the nullspace of its Laplacian matrix.
+
+-/
+
+
+open BigOperators Finset Matrix
+
+namespace SimpleGraph
+
+variable {V : Type*} (α : Type*)
+variable [Fintype V] [DecidableEq V] (G : SimpleGraph V) [DecidableRel G.Adj]
+
+/-- The diagonal matrix consisting of the degrees of the vertices in the graph. -/
+def degMatrix [AddMonoidWithOne α] : Matrix V V α := Matrix.diagonal (G.degree ·)
+
+/-- `lapMatrix G R` is the matrix `L = D - A` where `D`is the degree
+and `A` the adjacency matrix of `G`. -/
+def lapMatrix [AddGroupWithOne α] : Matrix V V α := G.degMatrix α - G.adjMatrix α
+
+variable {α}
+
+theorem isSymm_degMatrix [AddMonoidWithOne α] : (G.degMatrix α).IsSymm :=
+  isSymm_diagonal _
+
+theorem isSymm_lapMatrix [AddGroupWithOne α] : (G.lapMatrix α).IsSymm :=
+  (isSymm_degMatrix _).sub (isSymm_adjMatrix _)
+
+theorem degMatrix_mulVec_apply [NonAssocSemiring α] (v : V) (vec : V → α) :
+    (G.degMatrix α).mulVec vec v = G.degree v * vec v := by
+  rw [degMatrix, mulVec_diagonal]
+
+theorem lapMatrix_mulVec_apply [NonAssocRing α] (v : V) (vec : V → α) :
+    (G.lapMatrix α).mulVec vec v = G.degree v * vec v - ∑ u in G.neighborFinset v, vec u := by
+  simp_rw [lapMatrix, sub_mulVec, Pi.sub_apply, degMatrix_mulVec_apply, adjMatrix_mulVec_apply]
+
+theorem lapMatrix_mulVec_const_eq_zero [Ring α] : mulVec (G.lapMatrix α) (fun _ ↦ 1) = 0 := by
+  ext1 i
+  rw [lapMatrix_mulVec_apply]
+  simp
+
+theorem dotProduct_mulVec_adjMatrix [Ring α] (x : V → α) :
+    x ⬝ᵥ (G.adjMatrix α).mulVec x = ∑ i : V, ∑ j : V, if G.Adj i j then x i * x j else 0 := by
+  unfold mulVec dotProduct
+  simp [mul_sum]
```

Comment:
```suggestion
  simp only [dotProduct, mulVec, adjMatrix_apply, ite_mul, one_mul, zero_mul, mul_sum, mul_ite,
    mul_zero]
```

Reply:
Are `unfold`'s something that should be avoided in general?

Reply:
In the cases below I see why it makes sense but here using 'unfold' appears more readable to me.